project/src/main/java/com/google/sps/servlets/VolunteeringOpportunityDataServlet.java

```diff
@@ -27,19 +26,20 @@
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String opportunityId = request.getParameter(OPPORTUNITY_ID);
+    if (opportunityId == null) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "No opportunity id specified.");
+      return;
+    }
 
     Optional<VolunteeringOpportunity> opportunity =
         SpannerTasks.getVolunteeringOpportunityByOppportunityId(opportunityId);
-
-    if (opportunity.isPresent()) {
-      response.setContentType("application/json;");
-      response.getWriter().println(CommonUtils.convertToJson(opportunity.get()));
-    } else {
-      response.setContentType("text/html;");
+    if (!opportunity.isPresent()) {
```

Comment:
Flip logic here for readability

Reply:
Here I am going for the return early pattern.